### PR TITLE
1. Beaker tests for vxlan_global, 2. Fix for idempotence fail in the …

### PR DESCRIPTION
**Pull request by alok-aggarwal**
…provider file

1. Beaker tests passed, 
2. No rubocop errors

### CODE REVIEW

File: lib/puppet/provider/cisco_vxlan_global/nxapi.rb [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
   def initialize(value={})
     super(value)
-    @vxlan_global = Cisco::VxlanGlobal.new
+    @vxlan_global = Cisco::VxlanGlobal.new if Cisco::VxlanGlobal.enabled
```
Comment (mikewiebe) [inline lib/puppet/provider/cisco_vxlan_global/nxapi.rb]: What is the reason for the change?

File: tests/beaker_tests/vxlan_global/test_vxlan_global.rb [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+  manifest_props: '',
+  code:           [0, 2],
+  resource_props: {},
+}
+
+tests['default_properties'] = {
+  manifest_props: "
+    dup_host_ip_addr_detection_host_moves     => 'default',
+    dup_host_ip_addr_detection_timeout        => 'default',
+    anycast_gateway_mac                       => 'default',
```
Comment (mikewiebe) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: Does this mean that anycast_gateway_mac is not configured by default and is therefore not displayed in the `puppet resource` output?

File: lib/puppet/provider/cisco_vxlan_global/nxapi.rb (hunk 1 above)
Comment (alok-aggarwal) [inline lib/puppet/provider/cisco_vxlan_global/nxapi.rb]: ensure: absent Idempotence beaker test for default properties was always having to remove the resource. Thats why the test was failing with exit code 2, whereas the idempotence test expects always 0.

The reason was in provider initialize, a new instance was always getting created even the resource was disabled and therefore the ensure:absent always removed it which was incorrect. This check fixes it.

File: tests/beaker_tests/vxlan_global/test_vxlan_global.rb [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+# Full command string for puppet resource command
+def puppet_resource_cmd
+  cmd = UtilityLib::PUPPET_BINPATH + 'resource cisco_vxlan_global'
+  UtilityLib.get_namespace_cmd(agent, cmd, options)
+end
+
+def build_manifest_vxlan_global(tests, id)
+  if tests[id][:ensure] == :absent
+    state = 'ensure => absent,'
+    tests[id][:resource] = {}
```
Comment (mikewiebe) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: Should this be `tests[id][:resource] = { 'ensure' => 'absent' }` ?

File: tests/beaker_tests/vxlan_global/test_vxlan_global.rb (hunk 2 above)
Comment (alok-aggarwal) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: Thats correct. anycast_gateway_mac doesn't have a defualt value. The default means to unconfigure it.

File: lib/puppet/provider/cisco_vxlan_global/nxapi.rb (hunk 1 above)
Comment (chrisvanheuveln) [inline lib/puppet/provider/cisco_vxlan_global/nxapi.rb]: @mikewiebe We worked through this during consult hours. The problem is that the NU code will instantiate (by default) when you call VxlanGlobal.new; so self.instances gets a new instance every time, meaning ensure => :present is true; meaning destroy gets called every time, breaking idempotence.

File: tests/beaker_tests/vxlan_global/test_vxlan_global.rb (hunk 3 above)
Comment (alok-aggarwal) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: I think state = 'ensure => absent,' is doing the same.

Comment (chrisvanheuveln) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: @alok-aggarwal It's subtle but they're different. 
state is specifying the manifest entry, while resource specifies the puppet resource output (which needs 'absent' in quotes)

Comment (alok-aggarwal) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: well I used the template file where it is { }.

Adding tests[id][:resource] = { 'ensure' => 'absent' } gives me a failure.
Failed Tests Cases:
      Test Case vxlan_global/test_vxlan_global.rb reported: #<Beaker::DSL::Outcomes::FailTest: 
    TestStep :: Match (?-mix:ensure +=> +'?absent'?) :: FAIL
    >

Do i need to make a change elsewhere ?

Comment (mikewiebe) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: What output do you get from the `puppet resource cisco_vxlan_global` on your nx device when the resource is absent?

Comment (alok-aggarwal) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: Just blank.

root@agent-lab1-nx#puppet resource cisco_vxlan_global

root@agent-lab1-nx#

Comment (mikewiebe) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: @alok-aggarwal Ok, then that is the problem.  I think we can leave the code as `tests[id][:resource] = {}` but please gather the output of the beaker test with `--log-level trace` and make sure the test is issuing the following command `puppet resource cisco_vxlan_global` with no optional args.

Comment (alok-aggarwal) [inline tests/beaker_tests/vxlan_global/test_vxlan_global.rb]: yes it does.

**Comment by mikewiebe** [comment]
@alok-aggarwal Changes look good, but I just realized that you have not yet updated the changelog and readme file.  

* https://github.com/cisco/cisco-network-puppet-module/blob/feature/vxlan/CHANGELOG.md#new-feature-support
  * Add an entry under `New feature support`
* https://github.com/cisco/cisco-network-puppet-module/blob/feature/vxlan/README.md#resource-reference
  * Add an entry for your provider along with attributes under both `Resource Type Catalog (by Technology)` and `Cisco Resource Type Catalog (by Name)`

**Comment by alok-aggarwal** [comment]
Done.
